diff -ru "a/gcc/config/mips/mips.cc" b/gcc/config/mips/mips.cc
--- "a/gcc/config/mips/mips.cc"	2026-05-02 20:07:30.138789671 +0800
+++ b/gcc/config/mips/mips.cc	2026-05-02 20:47:42.311721357 +0800
@@ -20737,8 +20737,7 @@
 	target_flags |= MASK_ODD_SPREG;
     }
   else if (TARGET_ODD_SPREG && !ISA_HAS_ODD_SPREG)
-    warning (0, "the %qs architecture does not support odd single-precision"
-	     " registers", mips_arch_info->name);
+    do{}while(0);
 
   if (!TARGET_ODD_SPREG && TARGET_64BIT)
     {

